### DISCUSSION TITLE
Implementing dialects: practice of passing imformation between steps using kwargs?

### DISCUSSION THREAD

**Discussion by jimfulton** [sqlalchemy/sqlalchemy#6931] [Usage Questions] [answered]
Hi,

There seems to be a somewhat common practice of passing information between operation/steps using keyword arguments.

For example:
https://github.com/sqlalchemy/sqlalchemy/blob/207ec35c2e175f5fcf68e886d5e61a0678c2d6fc/lib/sqlalchemy/dialects/mysql/base.py#L2131
https://github.com/sqlalchemy/sqlalchemy/blob/207ec35c2e175f5fcf68e886d5e61a0678c2d6fc/lib/sqlalchemy/dialects/oracle/base.py#L931
https://github.com/sqlalchemy/sqlalchemy/blob/207ec35c2e175f5fcf68e886d5e61a0678c2d6fc/lib/sqlalchemy/sql/compiler.py#L2133-L2134

So, is this a common practice, or am I imagining it? :)

If so, is this documented anywhere?

I ask because I'm using this mechanism to pass information between `visit_table_valued_alias` and `visit_column` and a reviewer is reasonably asking if this is a standard practice.

**Comment by zzzeek** [comment, marked as answer]
first comment is that your first example is not the same as the second two.  the second two involve how the compiler's process() method has a `**kw` that will pass anything on through to the visit_ methods, which seems to be what you are doing also.   so to the question of, "is this **kw passed to process() and delivered on forward documented", I think the answer is "not much".  the kwargs can be set up initially from an argument called compile_kwargs that has a docstring, but the `.process()` method itself doesn't have docs as the `SQLCompiler` originally did not seek to "encourage" this method being used by outside dialects, because it was new and was also not consistently implemented for a long time, but also didn't exactly "discourage" it either as dialects like that of Oracle and SQL Server in particular make use of their own args.

So that said if you are using your own values in the `**kw` passed through compiler.process() and all the visit_ methods, that's fine, might be a good idea to qualify the names you use so they don't conflict with future names.

Then for the first example you have, that's a different "kw" system altogether and that one **is** documented, at least in terms of API docs with a little bit of example, at https://docs.sqlalchemy.org/en/14/core/sqlelement.html?highlight=dialectkwargs#sqlalchemy.sql.base.DialectKWArgs .   it's still stuff for dialect implementors only and does not typically see much use in the third party dialect world and is used for adding additional dialect-specific arguments mostly to schema-level DDLElement constructs.

**Reply by jimfulton** [reply]
> So that said if you are using your own values in the `**kw` passed through compiler.process() and all the visit_ methods, that's fine, 

Cool... I think.  I'm not passing anything through `process`.  I'm adding a key in `visit_table_valued_alias` 

https://github.com/googleapis/python-bigquery-sqlalchemy/blob/11ca126fe7aef0815557b7d2031f3e5ee18decd6/sqlalchemy_bigquery/base.py#L292

and using/updating it in `visit_column`.

https://github.com/googleapis/python-bigquery-sqlalchemy/blob/11ca126fe7aef0815557b7d2031f3e5ee18decd6/sqlalchemy_bigquery/base.py#L338-L343

There are no intermediate `process` calls.

> might be a good idea to qualify the names you use so they don't conflict with future names.

Yup. https://github.com/googleapis/python-bigquery-sqlalchemy/blob/11ca126fe7aef0815557b7d2031f3e5ee18decd6/sqlalchemy_bigquery/base.py#L65

**Reply by zzzeek** [reply]
> > So that said if you are using your own values in the `**kw` passed through compiler.process() and all the visit_ methods, that's fine,
> 
> Cool... I think. I'm not passing anything through `process`. I'm adding a key in `visit_table_valued_alias`
> 
> https://github.com/googleapis/python-bigquery-sqlalchemy/blob/11ca126fe7aef0815557b7d2031f3e5ee18decd6/sqlalchemy_bigquery/base.py#L292

OK "process()" is like a front end to any set of "visit" methods, it basically does the dispatch of getting the function named after "visit_name" from the given object and calling it.  so if you are passing kw into visit_xyz(), normally you would expect that visit_xyz() internally is going to call self.process() for all the internal bits of the element being passed.   as it turns out the SQLCompiler often skips the "process()" part but that's just to cut down on overhead.   

it's all good though, the general flow is process(an_XYZ, **kw) -> visit_XYZ(an_XYZ, **kw) -> process(elements of an_XYZ, **kw) -> (visit_abc(some_abc, **kw) / visit_def(some_def, **kw) / ...), it's mostly modeled after the AST visitors in the python std lib more or less.


> 
> and using/updating it in `visit_column`.
> 
> https://github.com/googleapis/python-bigquery-sqlalchemy/blob/11ca126fe7aef0815557b7d2031f3e5ee18decd6/sqlalchemy_bigquery/base.py#L338-L343
> 
> There are no intermediate `process` calls.
> 
> > might be a good idea to qualify the names you use so they don't conflict with future names.
> 
> Yup. https://github.com/googleapis/python-bigquery-sqlalchemy/blob/11ca126fe7aef0815557b7d2031f3e5ee18decd6/sqlalchemy_bigquery/base.py#L65

looks good
